components/script/dom/path2d.rs

```diff
@@ -59,20 +63,198 @@ impl Path2D {
     pub(crate) fn segments(&self) -> Vec<PathSegment> {
         self.path.borrow().clone()
     }
+
+    fn is_path_empty(&self) -> bool {
+        self.path.borrow().is_empty()
+    }
+
+    fn last_point(path: &[PathSegment]) -> Option<PathSegment> {
+        if path.is_empty() {
+            return None;
+        }
+
+        // already checked if path segments is empty above, so safe to unwrap at this step
+        let last_point = path.last().unwrap();
+
+        if let PathSegment::ClosePath = last_point {
+            // find first point for the last subpath
+            let first_point_index = match path
+                .iter()
+                .rev()
+                .skip(1)
+                .position(|segment| *segment == PathSegment::ClosePath)
+            {
+                Some(index) => index + 1,
+                None => 0,
+            };
+            let first_point = path.get(first_point_index)?;
+
+            if let PathSegment::ClosePath = first_point {
+                return None;
+            } else {
+                return Some(*first_point);
+            }
+        }
+
+        Some(*last_point)
+    }
+
+    fn add_path(&self, other: &Path2D, transform: &DOMMatrix2DInit) -> ErrorResult {
+        // Step 1. If the Path2D object path has no subpaths, then return.
+        if other.is_path_empty() {
+            return Ok(());
+        }
+
+        // Step 2 Let matrix be the result of creating a DOMMatrix from the 2D dictionary transform.
+        let matrix = dommatrix2dinit_to_matrix(transform)?;
+
+        // Step 3. If one or more of matrix's m11 element, m12 element, m21
+        // element, m22 element, m41 element, or m42 element are infinite or
+        // NaN, then return.
+        if !matrix.m11.is_finite() ||
+            !matrix.m12.is_finite() ||
+            !matrix.m21.is_finite() ||
+            !matrix.m22.is_finite() ||
+            !matrix.m31.is_finite() ||
+            !matrix.m32.is_finite()
+        {
+            return Ok(());
+        }
+
+        // Step 4. Create a copy of all the subpaths in path. Let c be this copy.
+        let mut c = other.segments();
+
+        // Step 5. Transform all the coordinates and lines in c by the transform matrix matrix.
+        let apply_matrix_transform = |x: f32, y: f32| -> Point2D<f64> {
```

Comment:
The matrix uses `f64` for its calculations, should we maybe change the `PathSegment` to contain f64 instead? I had to do lossy conversion from `f64` to `f32` after calculating the transformation.

Reply:
Probably yes, but not in this PR.